src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.template;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TextElement;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Represents a re-usable template that produces a formatted
+ * {@link Text.Builder}. Elements will be appended to the result builder in the
+ * order that they are specified in {@link #of(Object...)}.
+ *
+ * TODO: Serialization
+ */
+public class TextTemplate {
+
+    /**
+     * Empty representation of a {@link TextTemplate}. This is returned if the
+     * array supplied to {@link #of(Object...)} is empty.
+     */
+    public static final TextTemplate EMPTY = new TextTemplate();
+
+    final Object[] elements;
+
+    TextTemplate(Object... elements) {
+        this.elements = elements;
```

Comment:
Actually this **should** make a copy of some sort (preferably `ImmutableList.copyOf`), because otherwise you can still modify it from the outside by modifying the array given to the method below, which doesn't make much sense.